Introduce `clusternodes` collection for enhanced cluster node management

raararaara opened the pull request:
<!--  Thanks for sending a pull request! -->

**What this PR does / why we need it**:

This PR introduces the `clusternodes` collection to manage information about all currently alive nodes in the cluster. While the existing `leaderships` collection managed information about the leader node in the cluster mode, there was no mechanism to track follower nodes.
By extending the functionality of `leaderships`, this new implementation allows for comprehensive management of node information, including both leaders and followers. 

This change aims to improve the data structure related to cluster management by allowing for comprehensive tracking of all nodes in the cluster, not just the leaders. The introduction of the `clusternodes` collection will enable better scalability and management of the cluster's state.

**Which issue(s) this PR fixes**:

<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->

```release-note

```

**Additional documentation**:

<!--
This section can be blank if this pull request does not require a release note.

Please use the following format for linking documentation:
- [Usage]: <link>
- [Other doc]: <link>
-->

```docs

```

**Checklist**:

- [x] Added relevant tests or not required
- [ ] Addressed and resolved all CodeRabbit review comments
- [ ] Didn't break anything


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* New Features
  * Added ability to list active cluster nodes, showing current leader and followers for improved observability.

* Bug Fixes
  * Improved leader election stability under concurrency, reducing split-brain scenarios.
  * Faster and more reliable leader failover when a node stops or expires.
  * Automatic cleanup of inactive nodes to keep cluster status accurate.

* Tests
  * Added integration tests for concurrent startup, leader failover, and active node discovery.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

raararaara committed:
Introduce clusternodes to record followers and leader

raararaara committed:
Replace leaderships to clusternodes

raararaara committed:
Replace leaderships to clusternodes

raararaara committed:
Revise the codes

raararaara committed:
Fix follower upsert logic to ensure proper insertion

raararaara committed:
Add related tests

raararaara committed:
Merge branch 'main' into clusternodes

raararaara committed:
Add test for FindActiveClusterNodes

raararaara committed:
Fix follower upsert txn handling in memdb

raararaara committed:
Apply coderabbit's suggestion

hackerwins-yorkie reviewed (commented):
<!-- yorkie-ci-benchmark n8n comment -->
## Go Benchmark Analysis 📊

This is a comparison result between the previous(9ed272bf483acc494601777f9d90aa97e6f6c021) and the current commit(22f730e705c55111d978ccd4419edc6ea8014cf9).


### Significant Changes (≥20% difference)

| Benchmark suite | Previous | Current | Change |
|----------------|----------|---------|--------|
| BenchmarkVersionVector/clients_10/ (4_attach(ms)) | 5.00 ms | 4.00 ms | 🟢 -20.00% |
| BenchmarkVersionVector/clients_1000/ (3_pushpull(ms)) | 4.00 ms | 7.00 ms | 🔴 +75.00% |



### Key Observations 🔍
* Significant decrease of 20.00% in the `BenchmarkVersionVector` suite for the operation `4_attach(ms)` for `clients_10/` indicates an improvement in performance.
* A noticeable trend across various benchmarks is a slight improvement in performance, with multiple entries showing reductions in time or bytes on the order of 1-10%.

Overall, the benchmark data showcases some improvements in performance across different suites and operations.



### Clock Analysis

<details>
<summary>Lamport vs VersionVector</summary>

```markdown
#### BenchmarkVersionVector Test

##### Lamport (v0.5.2) 
| Metric                             | 10 clients    | 100 clients   | 1000 clients   |
| ---------------------------------- | ------------- | ------------- | -------------- |
| Total Operation Time               | 84.96 ms      | 793.94 ms     | 34.79 s        |
| Memory Allocations                 | 35.11 MB      | 219.92 MB     | 4.91 GB        |
| Number of Allocations (allocs/op)  | 69,271        | 1,246,728     | 81,485,288     |
| ChangePack Size                    | 138.0 B       | 137.0 B       | 141.0 B        |
| Snapshot Size                      | 379.0 B       | 3.08 KB       | 30.08 KB       |
| Push-Pull Time                     | 2.0 ms        | 1.5 ms        | 4.0 ms         |
| Attach Time                        | 4.5 ms        | 11.0 ms       | 31.0 ms        |
| ChangePack After Detach            | 138.0 B       | 140.0 B       | 141.0 B        |
| Snapshot After Detach              | 136.0 B       | 137.0 B       | 139.0 B        |
| Push-Pull After Detach             | 2.5 ms        | 5.0 ms        | 9.5 ms         |

##### Version Vector (current)
| Metric                             | 10 clients    | 100 clients   | 1000 clients   |
| ---------------------------------- | ------------- | ------------- | -------------- |
| Total Operation Time               | 114.83 ms     | 965.79 ms     | 11.86 s        |
| Memory Allocations                 | 8.22 MB       | 81.71 MB      | 1.46 GB        |
| Number of Allocations (allocs/op)  | 63,104        | 629,597       | 18,020,627     |
| ChangePack Size                    | 185.00 B      | 185.00 B      | 186.00 B       |
| Snapshot Size                      | 399.00 B      | 3.10 KB       | 30.10 KB       |
| Push-Pull Time                     | 3.00 ms       | 3.00 ms       | 7.00 ms        |
| Attach Time                        | 4.00 ms       | 6.00 ms       | 20.00 ms       |
| ChangePack After Detach            | 230.00 B      | 231.00 B      | 231.00 B       |
| Snapshot After Detach              | 156.00 B      | 156.00 B      | 156.00 B      |
| Push-Pull After Detach             | 4.00 ms       | 4.00 ms       | 4.00 ms        |

---

#### BenchmarkSyncConcurrency Test

| Metric                             | Clients          | Lamport (v0.5.2) | Version Vector (current) |
|-----------------------------------|------------------|------------------|---------------------------|
| Total Operation Time              | 1-100-10-10      | 7.48 s           | 8.03 s                  |
|                                   | 100-100-10-10    | 9.62 s           | 10.36 s                 |
|                                   | 300_100-10-10    | 14.77 s          | 12.37 s                 |
| Memory Allocations                | 1-100-10-10      | 1.15 GB          | 4.30 GB                 |
|                                   | 100-100-10-10    | 1.45 GB          | 4.42 GB                 |
|                                   | 300_100-10-10    | 2.19 GB          | 4.65 GB                 |
| Number of Allocations (allocs/op) | 1-100-10-10      | 17,107,766       | 106,386,105             |
|                                   | 100-100-10-10    | 20,084,480       | 107,685,573             |
|                                   | 300_100-10-10    | 30,359,215       | 110,252,660             |

---

#### Summary

- For the Version Vector implementation, we can see that as the number of clients increase, the Total Operation Time and Memory Allocations also increase, which is expected. However, the increase in the Version Vector Total Operation Time and Memory Allocations is relatively more pronounced compared to Lamport in a similar scenario.
  
- In terms of synchronization and concurrency, the Version Vector current implementation tends to have a slightly higher Total Operation Time and Memory Allocations compared to Lamport. This may signify that Version Vector requires more resources for synchronization tasks under a higher load of clients.

```

</details>



### Detailed Test Results

<details>
<summary>BenchmarkDeactivate</summary>

| Benchmark suite | Previous | Current | Change |
|----------------|----------|---------|--------|
| deactivate_with_100_total_50_attached/ (ns/op) | 142.18 ms | 138.53 ms | 🟢 -2.56% |
| deactivate_with_100_total_50_attached/ (B/op) | 13.06 MB | 12.78 MB | 🟢 -2.16% |
| deactivate_with_100_total_50_attached/ (allocs/op) | 153,173 allocs | 153,191 allocs | 🔴 +0.01% |
| deactivate_with_100_total_100_attached/ (ns/op) | 292.43 ms | 275.29 ms | 🟢 -5.86% |
| deactivate_with_100_total_100_attached/ (B/op) | 21.87 MB | 22.23 MB | 🔴 +1.64% |
| deactivate_with_100_total_100_attached/ (allocs/op) | 309,313 allocs | 308,827 allocs | 🟢 -0.16% |
</details>

<details>
<summary>BenchmarkDeletionConcurrency</summary>

| Benchmark suite | Previous | Current | Change |
|----------------|----------|---------|--------|
| concurrent_text_delete_range_100/ (ns/op) | 655.79 ms | 629.46 ms | 🟢 -4.02% |
| concurrent_text_delete_range_100/ (B/op) | 57.39 MB | 58.97 MB | 🔴 +2.75% |
| concurrent_text_delete_range_100/ (allocs/op) | 638,827 allocs | 638,980 allocs | 🔴 +0.02% |
| concurrent_text_delete_range_1000/ (ns/op) | 5.33 s | 5.13 s | 🟢 -3.63% |
| concurrent_text_delete_range_1000/ (B/op) | 342.23 MB | 351.82 MB | 🔴 +2.80% |
| concurrent_text_delete_range_1000/ (allocs/op) | 6,021,808 allocs | 6,023,463 allocs | 🔴 +0.03% |
| concurrent_tree_delete_range_100/ (ns/op) | 678.54 ms | 648.10 ms | 🟢 -4.49% |
| concurrent_tree_delete_range_100/ (B/op) | 64.93 MB | 65.07 MB | 🔴 +0.22% |
| concurrent_tree_delete_range_100/ (allocs/op) | 681,911 allocs | 681,826 allocs | 🟢 -0.01% |
| concurrent_tree_delete_range_1000/ (ns/op) | 6.17 s | 5.86 s | 🟢 -4.93% |
| concurrent_tree_delete_range_1000/ (B/op) | 879.61 MB | 885.55 MB | 🔴 +0.67% |
| concurrent_tree_delete_range_1000/ (allocs/op) | 6,487,132 allocs | 6,486,533 allocs | ⚪ 0% |
| concurrent_text_edit_delete_all_100/ (ns/op) | 578.78 ms | 553.24 ms | 🟢 -4.41% |
| concurrent_text_edit_delete_all_100/ (B/op) | 52.56 MB | 53.11 MB | 🔴 +1.05% |
| concurrent_text_edit_delete_all_100/ (allocs/op) | 540,589 allocs | 540,593 allocs | ⚪ 0% |
| concurrent_text_edit_delete_all_1000/ (ns/op) | 4.47 s | 4.29 s | 🟢 -4.01% |
| concurrent_text_edit_delete_all_1000/ (B/op) | 339.27 MB | 333.72 MB | 🟢 -1.63% |
| concurrent_text_edit_delete_all_1000/ (allocs/op) | 5,008,516 allocs | 5,008,619 allocs | ⚪ 0% |
| concurrent_tree_edit_delete_all_100/ (ns/op) | 584.41 ms | 558.29 ms | 🟢 -4.47% |
| concurrent_tree_edit_delete_all_100/ (B/op) | 58.83 MB | 57.69 MB | 🟢 -1.95% |
| concurrent_tree_edit_delete_all_100/ (allocs/op) | 587,340 allocs | 587,332 allocs | ⚪ 0% |
| concurrent_tree_edit_delete_all_1000/ (ns/op) | 5.10 s | 4.86 s | 🟢 -4.78% |
| concurrent_tree_edit_delete_all_1000/ (B/op) | 696.85 MB | 695.14 MB | 🟢 -0.25% |
| concurrent_tree_edit_delete_all_1000/ (allocs/op) | 5,365,535 allocs | 5,365,449 allocs | ⚪ 0% |
</details>

<details>
<summary>BenchmarkDocument</summary>

| Benchmark suite | Previous | Current | Change |
|----------------|----------|---------|--------|
| constructor_test/ (ns/op) | 1697.00 ns | 1573.00 ns | 🟢 -7.31% |
| constructor_test/ (B/op) | 1.59 KB | 1.59 KB | ⚪ 0% |
| constructor_test/ (allocs/op) | 27 allocs | 27 allocs | ⚪ 0% |
| status_test/ (ns/op) | 1280.00 ns | 1169.00 ns | 🟢 -8.67% |
| status_test/ (B/op) | 1.56 KB | 1.56 KB | ⚪ 0% |
| status_test/ (allocs/op) | 25 allocs | 25 allocs | ⚪ 0% |
| equals_test/ (ns/op) | 9142.00 ns | 8491.00 ns | 🟢 -7.12% |
| equals_test/ (B/op) | 8.33 KB | 8.33 KB | ⚪ 0% |
| equals_test/ (allocs/op) | 139 allocs | 139 allocs | ⚪ 0% |
| nested_update_test/ (ns/op) | 18837.00 ns | 17755.00 ns | 🟢 -5.74% |
| nested_update_test/ (B/op) | 12.86 KB | 12.86 KB | ⚪ 0% |
| nested_update_test/ (allocs/op) | 272 allocs | 272 allocs | ⚪ 0% |
| delete_test/ (ns/op) | 25200.00 ns | 23739.00 ns | 🟢 -5.80% |
| delete_test/ (B/op) | 16.37 KB | 16.37 KB | ⚪ 0% |
| delete_test/ (allocs/op) | 352 allocs | 352 allocs | ⚪ 0% |
| object_test/ (ns/op) | 9636.00 ns | 9135.00 ns | 🟢 -5.20% |
| object_test/ (B/op) | 7.42 KB | 7.42 KB | ⚪ 0% |
| object_test/ (allocs/op) | 124 allocs | 124 allocs | ⚪ 0% |
| array_test/ (ns/op) | 30685.00 ns | 29307.00 ns | 🟢 -4.49% |
| array_test/ (B/op) | 12.64 KB | 12.64 KB | ⚪ 0% |
| array_test/ (allocs/op) | 283 allocs | 283 allocs | ⚪ 0% |
| text_test/ (ns/op) | 34604.00 ns | 32859.00 ns | 🟢 -5.04% |
| text_test/ (B/op) | 15.31 KB | 15.31 KB | ⚪ 0% |
| text_test/ (allocs/op) | 503 allocs | 503 allocs | ⚪ 0% |
| text_composition_test/ (ns/op) | 33691.00 ns | 32027.00 ns | 🟢 -4.94% |
| text_composition_test/ (B/op) | 16.95 KB | 16.94 KB | ⚪ 0% |
| text_composition_test/ (allocs/op) | 493 allocs | 493 allocs | ⚪ 0% |
| rich_text_test/ (ns/op) | 90506.00 ns | 86752.00 ns | 🟢 -4.15% |
| rich_text_test/ (B/op) | 38.48 KB | 38.48 KB | ⚪ 0% |
| rich_text_test/ (allocs/op) | 1,185 allocs | 1,185 allocs | ⚪ 0% |
| counter_test/ (ns/op) | 19399.00 ns | 18107.00 ns | 🟢 -6.66% |
| counter_test/ (B/op) | 12.31 KB | 12.31 KB | ⚪ 0% |
| counter_test/ (allocs/op) | 254 allocs | 254 allocs | ⚪ 0% |
| text_edit_gc_100/ (ns/op) | 1.57 ms | 1.47 ms | 🟢 -6.25% |
| text_edit_gc_100/ (B/op) | 808.45 KB | 808.45 KB | ⚪ 0% |
| text_edit_gc_100/ (allocs/op) | 16,885 allocs | 16,885 allocs | ⚪ 0% |
| text_edit_gc_1000/ (ns/op) | 62.27 ms | 57.90 ms | 🟢 -7.01% |
| text_edit_gc_1000/ (B/op) | 46.27 MB | 46.27 MB | ⚪ 0% |
| text_edit_gc_1000/ (allocs/op) | 181,592 allocs | 181,593 allocs | ⚪ 0% |
| text_split_gc_100/ (ns/op) | 2.33 ms | 2.18 ms | 🟢 -6.59% |
| text_split_gc_100/ (B/op) | 1.53 MB | 1.53 MB | ⚪ 0% |
| text_split_gc_100/ (allocs/op) | 15,554 allocs | 15,553 allocs | ⚪ 0% |
| text_split_gc_1000/ (ns/op) | 141.35 ms | 132.55 ms | 🟢 -6.22% |
| text_split_gc_1000/ (B/op) | 137.28 MB | 137.28 MB | ⚪ 0% |
| text_split_gc_1000/ (allocs/op) | 180,994 allocs | 180,993 allocs | ⚪ 0% |
| text_100/ (ns/op) | 245752.00 ns | 236168.00 ns | 🟢 -3.90% |
| text_100/ (B/op) | 113.26 KB | 113.26 KB | ⚪ 0% |
| text_100/ (allocs/op) | 4,985 allocs | 4,985 allocs | ⚪ 0% |
| text_1000/ (ns/op) | 2.65 ms | 2.56 ms | 🟢 -3.39% |
| text_1000/ (B/op) | 1.08 MB | 1.08 MB | ⚪ 0% |
| text_1000/ (allocs/op) | 49,088 allocs | 49,088 allocs | ⚪ 0% |
| array_1000/ (ns/op) | 1.39 ms | 1.33 ms | 🟢 -3.89% |
| array_1000/ (B/op) | 1.13 MB | 1.13 MB | 🟢 -0.02% |
| array_1000/ (allocs/op) | 13,884 allocs | 13,883 allocs | ⚪ 0% |
| array_10000/ (ns/op) | 15.52 ms | 14.85 ms | 🟢 -4.33% |
| array_10000/ (B/op) | 10.29 MB | 10.29 MB | ⚪ 0% |
| array_10000/ (allocs/op) | 140,737 allocs | 140,737 allocs | ⚪ 0% |
| array_gc_100/ (ns/op) | 153445.00 ns | 142900.00 ns | 🟢 -6.87% |
| array_gc_100/ (B/op) | 104.26 KB | 104.28 KB | 🔴 +0.01% |
| array_gc_100/ (allocs/op) | 1,469 allocs | 1,469 allocs | ⚪ 0% |
| array_gc_1000/ (ns/op) | 1.61 ms | 1.53 ms | 🟢 -5.32% |
| array_gc_1000/ (B/op) | 1.18 MB | 1.18 MB | ⚪ 0% |
| array_gc_1000/ (allocs/op) | 14,929 allocs | 14,929 allocs | ⚪ 0% |
| counter_1000/ (ns/op) | 217814.00 ns | 202222.00 ns | 🟢 -7.16% |
| counter_1000/ (B/op) | 194.39 KB | 194.39 KB | ⚪ 0% |
| counter_1000/ (allocs/op) | 5,773 allocs | 5,773 allocs | ⚪ 0% |
| counter_10000/ (ns/op) | 2.27 ms | 2.21 ms | 🟢 -2.97% |
| counter_10000/ (B/op) | 2.23 MB | 2.23 MB | ⚪ 0% |
| counter_10000/ (allocs/op) | 59,780 allocs | 59,780 allocs | ⚪ 0% |
| object_1000/ (ns/op) | 1.63 ms | 1.55 ms | 🟢 -4.43% |
| object_1000/ (B/op) | 1.48 MB | 1.48 MB | ⚪ 0% |
| object_1000/ (allocs/op) | 11,926 allocs | 11,926 allocs | ⚪ 0% |
| object_10000/ (ns/op) | 17.64 ms | 17.25 ms | 🟢 -2.19% |
| object_10000/ (B/op) | 12.75 MB | 12.75 MB | 🔴 +0.02% |
| object_10000/ (allocs/op) | 121,233 allocs | 121,242 allocs | ⚪ 0% |
| tree_100/ (ns/op) | 786367.00 ns | 752594.00 ns | 🟢 -4.29% |
| tree_100/ (B/op) | 524.17 KB | 524.17 KB | ⚪ 0% |
| tree_100/ (allocs/op) | 4,722 allocs | 4,722 allocs | ⚪ 0% |
| tree_1000/ (ns/op) | 54.98 ms | 51.69 ms | 🟢 -5.99% |
| tree_1000/ (B/op) | 43.74 MB | 43.74 MB | ⚪ 0% |
| tree_1000/ (allocs/op) | 46,131 allocs | 46,131 allocs | ⚪ 0% |
| tree_10000/ (ns/op) | 7.06 s | 6.79 s | 🟢 -3.83% |
| tree_10000/ (B/op) | 4.30 GB | 4.30 GB | ⚪ 0% |
| tree_10000/ (allocs/op) | 460,224 allocs | 460,226 allocs | ⚪ 0% |
| tree_edit_gc_100/ (ns/op) | 2.87 ms | 2.70 ms | 🟢 -5.87% |
| tree_edit_gc_100/ (B/op) | 2.40 MB | 2.40 MB | ⚪ 0% |
| tree_edit_gc_100/ (allocs/op) | 11,668 allocs | 11,668 allocs | ⚪ 0% |
| tree_edit_gc_1000/ (ns/op) | 235.21 ms | 213.12 ms | 🟢 -9.39% |
| tree_edit_gc_1000/ (B/op) | 213.95 MB | 213.95 MB | ⚪ 0% |
| tree_edit_gc_1000/ (allocs/op) | 116,155 allocs | 116,156 allocs | ⚪ 0% |
| tree_split_gc_100/ (ns/op) | 2.25 ms | 2.06 ms | 🟢 -8.15% |
| tree_split_gc_100/ (B/op) | 1.61 MB | 1.61 MB | 🟢 -0.03% |
| tree_split_gc_100/ (allocs/op) | 9,606 allocs | 9,606 allocs | ⚪ 0% |
| tree_split_gc_1000/ (ns/op) | 164.72 ms | 149.49 ms | 🟢 -9.24% |
| tree_split_gc_1000/ (B/op) | 149.85 MB | 149.90 MB | 🔴 +0.03% |
| tree_split_gc_1000/ (allocs/op) | 107,759 allocs | 107,759 allocs | ⚪ 0% |
</details>

<details>
<summary>BenchmarkDocumentDeletion</summary>

| Benchmark suite | Previous | Current | Change |
|----------------|----------|---------|--------|
| single_text_delete_all_10000/ (ns/op) | 21.96 ms | 18.64 ms | 🟢 -15.11% |
| single_text_delete_all_10000/ (B/op) | 10.51 MB | 10.51 MB | 🟢 -0.02% |
| single_text_delete_all_10000/ (allocs/op) | 76,132 allocs | 76,128 allocs | ⚪ 0% |
| single_text_delete_all_100000/ (ns/op) | 358.17 ms | 293.97 ms | 🟢 -17.93% |
| single_text_delete_all_100000/ (B/op) | 104.78 MB | 104.75 MB | 🟢 -0.03% |
| single_text_delete_all_100000/ (allocs/op) | 766,159 allocs | 766,138 allocs | ⚪ 0% |
| single_tree_delete_all_1000/ (ns/op) | 54.59 ms | 56.09 ms | 🔴 +2.74% |
| single_tree_delete_all_1000/ (B/op) | 44.66 MB | 44.66 MB | ⚪ 0% |
| single_tree_delete_all_1000/ (allocs/op) | 58,199 allocs | 58,204 allocs | ⚪ 0% |
| single_text_delete_range_100/ (ns/op) | 464533.00 ns | 474056.00 ns | 🔴 +2.05% |
| single_text_delete_range_100/ (B/op) | 243.26 KB | 243.27 KB | ⚪ 0% |
| single_text_delete_range_100/ (allocs/op) | 7,074 allocs | 7,074 allocs | ⚪ 0% |
| single_text_delete_range_1000/ (ns/op) | 9.60 ms | 9.41 ms | 🟢 -1.96% |
| single_text_delete_range_1000/ (B/op) | 6.37 MB | 6.37 MB | ⚪ 0% |
| single_text_delete_range_1000/ (allocs/op) | 73,369 allocs | 73,370 allocs | ⚪ 0% |
| single_tree_delete_range_100/ (ns/op) | 994725.00 ns | 1.00 ms | 🔴 +0.59% |
| single_tree_delete_range_100/ (B/op) | 708.46 KB | 708.39 KB | ⚪ 0% |
| single_tree_delete_range_100/ (allocs/op) | 6,392 allocs | 6,392 allocs | ⚪ 0% |
| single_tree_delete_range_1000/ (ns/op) | 63.42 ms | 64.83 ms | 🔴 +2.23% |
| single_tree_delete_range_1000/ (B/op) | 54.51 MB | 54.51 MB | ⚪ 0% |
| single_tree_delete_range_1000/ (allocs/op) | 63,431 allocs | 63,432 allocs | ⚪ 0% |
</details>

<details>
<summary>BenchmarkGetDocuments</summary>

| Benchmark suite | Previous | Current | Change |
|----------------|----------|---------|--------|
| without_root_presence_10/ (ns/op) | 1.15 ms | 1.14 ms | 🟢 -1.51% |
| without_root_presence_10/ (B/op) | 118.53 KB | 116.95 KB | 🟢 -1.33% |
| without_root_presence_10/ (allocs/op) | 1,015 allocs | 1,015 allocs | ⚪ 0% |
| with_root_presence_10/ (ns/op) | 4.06 ms | 3.96 ms | 🟢 -2.25% |
| with_root_presence_10/ (B/op) | 601.74 KB | 591.31 KB | 🟢 -1.73% |
| with_root_presence_10/ (allocs/op) | 5,615 allocs | 5,620 allocs | 🔴 +0.09% |
| without_root_presence_100/ (ns/op) | 2.41 ms | 2.41 ms | 🟢 -0.08% |
| without_root_presence_100/ (B/op) | 798.03 KB | 811.86 KB | 🔴 +1.73% |
| without_root_presence_100/ (allocs/op) | 5,947 allocs | 5,934 allocs | 🟢 -0.22% |
| with_root_presence_100/ (ns/op) | 43.37 ms | 42.51 ms | 🟢 -1.97% |
| with_root_presence_100/ (B/op) | 7.75 MB | 8.67 MB | 🔴 +11.87% |
| with_root_presence_100/ (allocs/op) | 67,111 allocs | 72,553 allocs | 🔴 +8.11% |
| without_root_presence_1000/ (ns/op) | 15.84 ms | 14.89 ms | 🟢 -6.00% |
| without_root_presence_1000/ (B/op) | 7.96 MB | 7.87 MB | 🟢 -1.14% |
| without_root_presence_1000/ (allocs/op) | 54,684 allocs | 54,676 allocs | 🟢 -0.01% |
| with_root_presence_1000/ (ns/op) | 414.56 ms | 405.28 ms | 🟢 -2.24% |
| with_root_presence_1000/ (B/op) | 72.54 MB | 77.65 MB | 🔴 +7.05% |
| with_root_presence_1000/ (allocs/op) | 662,142 allocs | 667,394 allocs | 🔴 +0.79% |
</details>

<details>
<summary>BenchmarkRPC</summary>

| Benchmark suite | Previous | Current | Change |
|----------------|----------|---------|--------|
| client_to_server/ (ns/op) | 228.74 ms | 220.58 ms | 🟢 -3.57% |
| client_to_server/ (B/op) | 14.99 MB | 13.86 MB | 🟢 -7.53% |
| client_to_server/ (allocs/op) | 176,792 allocs | 176,867 allocs | 🔴 +0.04% |
| client_to_client_via_server/ (ns/op) | 170.84 ms | 172.13 ms | 🔴 +0.76% |
| client_to_client_via_server/ (B/op) | 22.37 MB | 22.56 MB | 🔴 +0.82% |
| client_to_client_via_server/ (allocs/op) | 268,206 allocs | 268,733 allocs | 🔴 +0.20% |
| attach_large_document/ (ns/op) | 1.29 s | 1.28 s | 🟢 -0.95% |
| attach_large_document/ (B/op) | 1.85 GB | 1.85 GB | 🟢 -0.04% |
| attach_large_document/ (allocs/op) | 11,106 allocs | 11,174 allocs | 🔴 +0.61% |
| adminCli_to_server/ (ns/op) | 591.24 ms | 593.11 ms | 🔴 +0.32% |
| adminCli_to_server/ (B/op) | 36.18 MB | 34.99 MB | 🟢 -3.31% |
| adminCli_to_server/ (allocs/op) | 404,256 allocs | 404,231 allocs | ⚪ 0% |
</details>

<details>
<summary>BenchmarkLocker</summary>

| Benchmark suite | Previous | Current | Change |
|----------------|----------|---------|--------|
|  (ns/op) | 86.21 ns | 83.12 ns | 🟢 -3.58% |
|  (B/op) | 32.00 B | 32.00 B | ⚪ 0% |
|  (allocs/op) | 1 allocs | 1 allocs | ⚪ 0% |
</details>

<details>
<summary>BenchmarkLockerParallel</summary>

| Benchmark suite | Previous | Current | Change |
|----------------|----------|---------|--------|
|  (ns/op) | 46.53 ns | 46.31 ns | 🟢 -0.47% |
|  (B/op) | 0.00 B | 0.00 B | ⚪ 0% |
|  (allocs/op) | 0 allocs | 0 allocs | ⚪ 0% |
</details>

<details>
<summary>BenchmarkLockerMoreKeys</summary>

| Benchmark suite | Previous | Current | Change |
|----------------|----------|---------|--------|
|  (ns/op) | 188.90 ns | 184.70 ns | 🟢 -2.22% |
|  (B/op) | 31.00 B | 30.00 B | 🟢 -3.23% |
|  (allocs/op) | 0 allocs | 0 allocs | ⚪ 0% |
</details>

<details>
<summary>BenchmarkRWLocker</summary>

| Benchmark suite | Previous | Current | Change |
|----------------|----------|---------|--------|
| RWLock_rate_2/ (ns/op) | 50.75 ns | 50.04 ns | 🟢 -1.40% |
| RWLock_rate_2/ (B/op) | 0.00 B | 0.00 B | ⚪ 0% |
| RWLock_rate_2/ (allocs/op) | 0 allocs | 0 allocs | ⚪ 0% |
| RWLock_rate_10/ (ns/op) | 44.79 ns | 44.48 ns | 🟢 -0.69% |
| RWLock_rate_10/ (B/op) | 0.00 B | 0.00 B | ⚪ 0% |
| RWLock_rate_10/ (allocs/op) | 0 allocs | 0 allocs | ⚪ 0% |
| RWLock_rate_100/ (ns/op) | 61.71 ns | 61.57 ns | 🟢 -0.23% |
| RWLock_rate_100/ (B/op) | 2.00 B | 2.00 B | ⚪ 0% |
| RWLock_rate_100/ (allocs/op) | 0 allocs | 0 allocs | ⚪ 0% |
| RWLock_rate_1000/ (ns/op) | 92.34 ns | 88.98 ns | 🟢 -3.64% |
| RWLock_rate_1000/ (B/op) | 9.00 B | 8.00 B | 🟢 -11.11% |
| RWLock_rate_1000/ (allocs/op) | 0 allocs | 0 allocs | ⚪ 0% |
</details>

<details>
<summary>BenchmarkPresenceConcurrency</summary>

| Benchmark suite | Previous | Current | Change |
|----------------|----------|---------|--------|
| 0-100-10/ (ns/op) | 919.19 ms | 907.85 ms | 🟢 -1.23% |
| 0-100-10/ (B/op) | 407.76 MB | 405.79 MB | 🟢 -0.48% |
| 0-100-10/ (allocs/op) | 5,118,488 allocs | 5,119,066 allocs | 🔴 +0.01% |
| 100-100-10/ (ns/op) | 982.77 ms | 939.21 ms | 🟢 -4.43% |
| 100-100-10/ (B/op) | 446.30 MB | 428.28 MB | 🟢 -4.04% |
| 100-100-10/ (allocs/op) | 5,603,838 allocs | 5,603,210 allocs | 🟢 -0.01% |
| 300-100-10/ (ns/op) | 1.04 s | 1.01 s | 🟢 -3.53% |
| 300-100-10/ (B/op) | 495.85 MB | 497.60 MB | 🔴 +0.35% |
| 300-100-10/ (allocs/op) | 6,580,527 allocs | 6,577,811 allocs | 🟢 -0.04% |
</details>

<details>
<summary>BenchmarkChange</summary>

| Benchmark suite | Previous | Current | Change |
|----------------|----------|---------|--------|
| Push_10_Changes/ (ns/op) | 4.69 ms | 4.57 ms | 🟢 -2.50% |
| Push_10_Changes/ (B/op) | 348.92 KB | 347.88 KB | 🟢 -0.30% |
| Push_10_Changes/ (allocs/op) | 5,888 allocs | 5,862 allocs | 🟢 -0.44% |
| Push_100_Changes/ (ns/op) | 18.56 ms | 18.26 ms | 🟢 -1.60% |
| Push_100_Changes/ (B/op) | 1.60 MB | 1.60 MB | 🟢 -0.25% |
| Push_100_Changes/ (allocs/op) | 17,561 allocs | 17,512 allocs | 🟢 -0.28% |
| Push_1000_Changes/ (ns/op) | 151.10 ms | 149.43 ms | 🟢 -1.11% |
| Push_1000_Changes/ (B/op) | 13.68 MB | 13.79 MB | 🔴 +0.83% |
| Push_1000_Changes/ (allocs/op) | 108,495 allocs | 108,449 allocs | 🟢 -0.04% |
| Pull_10_Changes/ (ns/op) | 3.11 ms | 3.06 ms | 🟢 -1.37% |
| Pull_10_Changes/ (B/op) | 341.10 KB | 346.66 KB | 🔴 +1.63% |
| Pull_10_Changes/ (allocs/op) | 6,934 allocs | 7,102 allocs | 🔴 +2.42% |
| Pull_100_Changes/ (ns/op) | 5.04 ms | 4.92 ms | 🟢 -2.40% |
| Pull_100_Changes/ (B/op) | 1.14 MB | 1.16 MB | 🔴 +1.86% |
| Pull_100_Changes/ (allocs/op) | 18,697 allocs | 19,132 allocs | 🔴 +2.33% |
| Pull_1000_Changes/ (ns/op) | 13.18 ms | 13.56 ms | 🔴 +2.90% |
| Pull_1000_Changes/ (B/op) | 6.72 MB | 6.73 MB | 🔴 +0.18% |
| Pull_1000_Changes/ (allocs/op) | 68,366 allocs | 68,890 allocs | 🔴 +0.77% |
</details>

<details>
<summary>BenchmarkSnapshot</summary>

| Benchmark suite | Previous | Current | Change |
|----------------|----------|---------|--------|
| Push_3KB_snapshot/ (ns/op) | 21.20 ms | 20.78 ms | 🟢 -1.95% |
| Push_3KB_snapshot/ (B/op) | 2.40 MB | 2.46 MB | 🔴 +2.49% |
| Push_3KB_snapshot/ (allocs/op) | 39,569 allocs | 40,041 allocs | 🔴 +1.19% |
| Push_30KB_snapshot/ (ns/op) | 158.59 ms | 158.07 ms | 🟢 -0.33% |
| Push_30KB_snapshot/ (B/op) | 18.02 MB | 18.01 MB | 🟢 -0.07% |
| Push_30KB_snapshot/ (allocs/op) | 190,607 allocs | 191,589 allocs | 🔴 +0.52% |
| Pull_3KB_snapshot/ (ns/op) | 6.29 ms | 6.38 ms | 🔴 +1.35% |
| Pull_3KB_snapshot/ (B/op) | 1.82 MB | 1.84 MB | 🔴 +0.64% |
| Pull_3KB_snapshot/ (allocs/op) | 34,519 allocs | 34,982 allocs | 🔴 +1.34% |
| Pull_30KB_snapshot/ (ns/op) | 18.80 ms | 17.87 ms | 🟢 -4.93% |
| Pull_30KB_snapshot/ (B/op) | 11.53 MB | 11.53 MB | 🔴 +0.07% |
| Pull_30KB_snapshot/ (allocs/op) | 164,168 allocs | 164,596 allocs | 🔴 +0.26% |
</details>

<details>
<summary>BenchmarkSplayTree</summary>

| Benchmark suite | Previous | Current | Change |
|----------------|----------|---------|--------|
| stress_test_100000/ (ns/op) | 0.18 ns | 0.17 ns | 🟢 -4.54% |
| stress_test_100000/ (B/op) | 0.00 B | 0.00 B | ⚪ 0% |
| stress_test_100000/ (allocs/op) | 0 allocs | 0 allocs | ⚪ 0% |
| stress_test_200000/ (ns/op) | 0.38 ns | 0.36 ns | 🟢 -3.73% |
| stress_test_200000/ (B/op) | 0.00 B | 0.00 B | ⚪ 0% |
| stress_test_200000/ (allocs/op) | 0 allocs | 0 allocs | ⚪ 0% |
| stress_test_300000/ (ns/op) | 0.54 ns | 0.55 ns | 🔴 +1.76% |
| stress_test_300000/ (B/op) | 0.00 B | 0.00 B | ⚪ 0% |
| stress_test_300000/ (allocs/op) | 0 allocs | 0 allocs | ⚪ 0% |
| random_access_100000/ (ns/op) | 0.01 ns | 0.01 ns | 🟢 -1.52% |
| random_access_100000/ (B/op) | 0.00 B | 0.00 B | ⚪ 0% |
| random_access_100000/ (allocs/op) | 0 allocs | 0 allocs | ⚪ 0% |
| random_access_200000/ (ns/op) | 0.03 ns | 0.04 ns | 🔴 +1.38% |
| random_access_200000/ (B/op) | 0.00 B | 0.00 B | ⚪ 0% |
| random_access_200000/ (allocs/op) | 0 allocs | 0 allocs | ⚪ 0% |
| random_access_300000/ (ns/op) | 0.05 ns | 0.04 ns | 🟢 -5.04% |
| random_access_300000/ (B/op) | 0.00 B | 0.00 B | ⚪ 0% |
| random_access_300000/ (allocs/op) | 0 allocs | 0 allocs | ⚪ 0% |
| editing_trace_bench/ (ns/op) | 0.00 ns | 0.00 ns | 🔴 +14.36% |
| editing_trace_bench/ (B/op) | 0.00 B | 0.00 B | ⚪ 0% |
| editing_trace_bench/ (allocs/op) | 0 allocs | 0 allocs | ⚪ 0% |
</details>

<details>
<summary>BenchmarkSync</summary>

| Benchmark suite | Previous | Current | Change |
|----------------|----------|---------|--------|
| memory_sync_10_test/ (ns/op) | 6885.00 ns | 7079.00 ns | 🔴 +2.82% |
| memory_sync_10_test/ (B/op) | 1.34 KB | 1.34 KB | ⚪ 0% |
| memory_sync_10_test/ (allocs/op) | 35 allocs | 35 allocs | ⚪ 0% |
| memory_sync_100_test/ (ns/op) | 55230.00 ns | 55796.00 ns | 🔴 +1.02% |
| memory_sync_100_test/ (B/op) | 9.51 KB | 9.52 KB | 🔴 +0.11% |
| memory_sync_100_test/ (allocs/op) | 268 allocs | 268 allocs | ⚪ 0% |
| memory_sync_1000_test/ (ns/op) | 606108.00 ns | 595125.00 ns | 🟢 -1.81% |
| memory_sync_1000_test/ (B/op) | 76.19 KB | 76.47 KB | 🔴 +0.36% |
| memory_sync_1000_test/ (allocs/op) | 2,122 allocs | 2,131 allocs | 🔴 +0.42% |
| memory_sync_10000_test/ (ns/op) | 8.07 ms | 7.45 ms | 🟢 -7.73% |
| memory_sync_10000_test/ (B/op) | 755.85 KB | 767.80 KB | 🔴 +1.58% |
| memory_sync_10000_test/ (allocs/op) | 20,440 allocs | 20,640 allocs | 🔴 +0.98% |
</details>

<details>
<summary>BenchmarkSyncConcurrency</summary>

| Benchmark suite | Previous | Current | Change |
|----------------|----------|---------|--------|
| 1-100-10/ (ns/op) | 8.01 s | 8.03 s | 🔴 +0.23% |
| 1-100-10/ (B/op) | 4.30 GB | 4.30 GB | 🟢 -0.02% |
| 1-100-10/ (allocs/op) | 106,408,583 allocs | 106,386,105 allocs | 🟢 -0.02% |
| 100-100-10/ (ns/op) | 10.63 s | 10.36 s | 🟢 -2.51% |
| 100-100-10/ (B/op) | 4.43 GB | 4.42 GB | 🟢 -0.27% |
| 100-100-10/ (allocs/op) | 107,551,635 allocs | 107,685,573 allocs | 🔴 +0.12% |
| 300_100-10/ (ns/op) | 12.83 s | 12.37 s | 🟢 -3.56% |
| 300_100-10/ (B/op) | 4.66 GB | 4.65 GB | 🟢 -0.10% |
| 300_100-10/ (allocs/op) | 110,440,582 allocs | 110,252,660 allocs | 🟢 -0.17% |
</details>

<details>
<summary>BenchmarkTextEditing</summary>

| Benchmark suite | Previous | Current | Change |
|----------------|----------|---------|--------|
|  (ns/op) | 5.76 s | 5.95 s | 🔴 +3.35% |
|  (B/op) | 3.88 GB | 3.88 GB | ⚪ 0% |
|  (allocs/op) | 20,100,372 allocs | 20,100,502 allocs | ⚪ 0% |
</details>

<details>
<summary>BenchmarkTree</summary>

| Benchmark suite | Previous | Current | Change |
|----------------|----------|---------|--------|
| 10000_vertices_to_protobuf/ (ns/op) | 4.82 ms | 4.79 ms | 🟢 -0.66% |
| 10000_vertices_to_protobuf/ (B/op) | 6.68 MB | 6.68 MB | ⚪ 0% |
| 10000_vertices_to_protobuf/ (allocs/op) | 90,026 allocs | 90,026 allocs | ⚪ 0% |
| 10000_vertices_from_protobuf/ (ns/op) | 232.38 ms | 230.02 ms | 🟢 -1.01% |
| 10000_vertices_from_protobuf/ (B/op) | 442.15 MB | 442.15 MB | ⚪ 0% |
| 10000_vertices_from_protobuf/ (allocs/op) | 280,047 allocs | 280,048 allocs | ⚪ 0% |
| 20000_vertices_to_protobuf/ (ns/op) | 10.73 ms | 10.30 ms | 🟢 -3.99% |
| 20000_vertices_to_protobuf/ (B/op) | 13.53 MB | 13.53 MB | ⚪ 0% |
| 20000_vertices_to_protobuf/ (allocs/op) | 180,029 allocs | 180,030 allocs | ⚪ 0% |
| 20000_vertices_from_protobuf/ (ns/op) | 896.40 ms | 892.57 ms | 🟢 -0.43% |
| 20000_vertices_from_protobuf/ (B/op) | 1.70 GB | 1.70 GB | ⚪ 0% |
| 20000_vertices_from_protobuf/ (allocs/op) | 560,080 allocs | 560,079 allocs | ⚪ 0% |
| 30000_vertices_to_protobuf/ (ns/op) | 15.10 ms | 15.11 ms | 🔴 +0.06% |
| 30000_vertices_to_protobuf/ (B/op) | 19.94 MB | 19.94 MB | ⚪ 0% |
| 30000_vertices_to_protobuf/ (allocs/op) | 270,030 allocs | 270,030 allocs | ⚪ 0% |
| 30000_vertices_from_protobuf/ (ns/op) | 1.98 s | 2.04 s | 🔴 +2.71% |
| 30000_vertices_from_protobuf/ (B/op) | 3.75 GB | 3.75 GB | ⚪ 0% |
| 30000_vertices_from_protobuf/ (allocs/op) | 840,234 allocs | 840,129 allocs | 🟢 -0.01% |
</details>

<details>
<summary>BenchmarkVersionVector</summary>

| Benchmark suite | Previous | Current | Change |
|----------------|----------|---------|--------|
| clients_10/ (ns/op) | 114.87 ms | 114.83 ms | 🟢 -0.04% |
| clients_10/ (1_changepack(bytes)) | 185.00 B | 185.00 B | ⚪ 0% |
| clients_10/ (2_snapshot(bytes)) | 399.00 B | 399.00 B | ⚪ 0% |
| clients_10/ (3_pushpull(ms)) | 3.00 ms | 3.00 ms | ⚪ 0% |
| clients_10/ (4_attach(ms)) | 5.00 ms | 4.00 ms | 🟢 -20.00% |
| clients_10/ (5_changepack_after_detach(bytes)) | 230.00 B | 230.00 B | ⚪ 0% |
| clients_10/ (6_snapshot_after_detach(bytes)) | 156.00 B | 156.00 B | ⚪ 0% |
| clients_10/ (7_pushpull_after_detach(ms)) | 4.00 ms | 4.00 ms | ⚪ 0% |
| clients_10/ (B/op) | 9.22 MB | 8.22 MB | 🟢 -10.88% |
| clients_10/ (allocs/op) | 63,134 allocs | 63,104 allocs | 🟢 -0.05% |
| clients_100/ (ns/op) | 961.08 ms | 965.79 ms | 🔴 +0.49% |
| clients_100/ (1_changepack(bytes)) | 185.00 B | 185.00 B | ⚪ 0% |
| clients_100/ (2_snapshot(bytes)) | 3.10 KB | 3.10 KB | ⚪ 0% |
| clients_100/ (3_pushpull(ms)) | 3.00 ms | 3.00 ms | ⚪ 0% |
| clients_100/ (4_attach(ms)) | 6.00 ms | 6.00 ms | ⚪ 0% |
| clients_100/ (5_changepack_after_detach(bytes)) | 231.00 B | 231.00 B | ⚪ 0% |
| clients_100/ (6_snapshot_after_detach(bytes)) | 156.00 B | 156.00 B | ⚪ 0% |
| clients_100/ (7_pushpull_after_detach(ms)) | 4.00 ms | 4.00 ms | ⚪ 0% |
| clients_100/ (B/op) | 76.05 MB | 81.71 MB | 🔴 +7.44% |
| clients_100/ (allocs/op) | 629,434 allocs | 629,597 allocs | 🔴 +0.03% |
| clients_1000/ (ns/op) | 11.93 s | 11.86 s | 🟢 -0.60% |
| clients_1000/ (1_changepack(bytes)) | 186.00 B | 186.00 B | ⚪ 0% |
| clients_1000/ (2_snapshot(bytes)) | 30.10 KB | 30.10 KB | ⚪ 0% |
| clients_1000/ (3_pushpull(ms)) | 4.00 ms | 7.00 ms | 🔴 +75.00% |
| clients_1000/ (4_attach(ms)) | 22.00 ms | 20.00 ms | 🟢 -9.09% |
| clients_1000/ (5_changepack_after_detach(bytes)) | 231.00 B | 231.00 B | ⚪ 0% |
| clients_1000/ (6_snapshot_after_detach(bytes)) | 156.00 B | 156.00 B | ⚪ 0% |
| clients_1000/ (7_pushpull_after_detach(ms)) | 4.00 ms | 4.00 ms | ⚪ 0% |
| clients_1000/ (B/op) | 1.46 GB | 1.46 GB | 🟢 -0.06% |
| clients_1000/ (allocs/op) | 18,021,108 allocs | 18,020,627 allocs | ⚪ 0% |
</details>

<details>
<summary>BenchmarkWebhook</summary>

| Benchmark suite | Previous | Current | Change |
|----------------|----------|---------|--------|
| Send_10_Webhooks_to_10_Endpoints/ (ns/op) | 12.27 ms | 12.40 ms | 🔴 +1.11% |
| Send_10_Webhooks_to_10_Endpoints/ (B/op) | 808.37 KB | 808.46 KB | 🔴 +0.01% |
| Send_10_Webhooks_to_10_Endpoints/ (allocs/op) | 10,116 allocs | 10,116 allocs | ⚪ 0% |
| Send_100_Webhooks_to_10_Endpoints/ (ns/op) | 122.65 ms | 122.96 ms | 🔴 +0.26% |
| Send_100_Webhooks_to_10_Endpoints/ (B/op) | 8.08 MB | 8.08 MB | 🟢 -0.01% |
| Send_100_Webhooks_to_10_Endpoints/ (allocs/op) | 101,166 allocs | 101,153 allocs | 🟢 -0.01% |
| Send_10_Webhooks_to_100_Endpoints/ (ns/op) | 125.64 ms | 125.20 ms | 🟢 -0.35% |
| Send_10_Webhooks_to_100_Endpoints/ (B/op) | 8.28 MB | 8.28 MB | 🟢 -0.03% |
| Send_10_Webhooks_to_100_Endpoints/ (allocs/op) | 102,538 allocs | 102,531 allocs | ⚪ 0% |
| Send_100_Webhooks_to_100_Endpoints/ (ns/op) | 1.24 s | 1.25 s | 🔴 +1.23% |
| Send_100_Webhooks_to_100_Endpoints/ (B/op) | 82.36 MB | 82.37 MB | 🔴 +0.02% |
| Send_100_Webhooks_to_100_Endpoints/ (allocs/op) | 1,022,470 allocs | 1,022,460 allocs | ⚪ 0% |
| Send_10_Webhooks_to_1000_Endpoints/ (ns/op) | 2.76 s | 2.77 s | 🔴 +0.51% |
| Send_10_Webhooks_to_1000_Endpoints/ (B/op) | 209.72 MB | 209.64 MB | 🟢 -0.04% |
| Send_10_Webhooks_to_1000_Endpoints/ (allocs/op) | 1,716,687 allocs | 1,716,772 allocs | ⚪ 0% |
</details>

<details>
<summary>BenchmarkWebhookWithLimit</summary>

| Benchmark suite | Previous | Current | Change |
|----------------|----------|---------|--------|
| Send_10_Webhooks_to_10_Endpoints_with_limit/ (ns/op) | 3.80 ms | 3.77 ms | 🟢 -0.83% |
| Send_10_Webhooks_to_10_Endpoints_with_limit/ (B/op) | 306.43 KB | 306.40 KB | 🟢 -0.01% |
| Send_10_Webhooks_to_10_Endpoints_with_limit/ (allocs/op) | 2,913 allocs | 2,913 allocs | ⚪ 0% |
| Send_100_Webhooks_to_10_Endpoints_with_limit/ (ns/op) | 4.04 ms | 4.04 ms | 🔴 +0.03% |
| Send_100_Webhooks_to_10_Endpoints_with_limit/ (B/op) | 428.88 KB | 428.87 KB | ⚪ 0% |
| Send_100_Webhooks_to_10_Endpoints_with_limit/ (allocs/op) | 4,713 allocs | 4,714 allocs | 🔴 +0.02% |
| Send_10_Webhooks_to_100_Endpoints_with_limit/ (ns/op) | 37.45 ms | 37.78 ms | 🔴 +0.91% |
| Send_10_Webhooks_to_100_Endpoints_with_limit/ (B/op) | 3.08 MB | 3.08 MB | 🔴 +0.04% |
| Send_10_Webhooks_to_100_Endpoints_with_limit/ (allocs/op) | 29,167 allocs | 29,175 allocs | 🔴 +0.03% |
| Send_100_Webhooks_to_100_Endpoints_with_limit/ (ns/op) | 40.37 ms | 40.65 ms | 🔴 +0.71% |
| Send_100_Webhooks_to_100_Endpoints_with_limit/ (B/op) | 4.37 MB | 4.37 MB | ⚪ 0% |
| Send_100_Webhooks_to_100_Endpoints_with_limit/ (allocs/op) | 47,200 allocs | 47,197 allocs | ⚪ 0% |
| Send_10_Webhooks_to_1000_Endpoints_with_limit/ (ns/op) | 367.03 ms | 371.53 ms | 🔴 +1.23% |
| Send_10_Webhooks_to_1000_Endpoints_with_limit/ (B/op) | 44.53 MB | 44.54 MB | 🔴 +0.03% |
| Send_10_Webhooks_to_1000_Endpoints_with_limit/ (allocs/op) | 380,169 allocs | 380,183 allocs | ⚪ 0% |
</details>

hackerwins-yorkie reviewed (commented):
<!-- yorkie-ci-load n8n comment -->
## k6 Load Test 📊

This is a comparison result between the previous(9ed272bf483acc494601777f9d90aa97e6f6c021) and the current commit(22f730e705c55111d978ccd4419edc6ea8014cf9).


### Change

| Metric | Previous | Current | Change |
|-----|------|------|--------|
| transaction_time.avg | 6611.54 | 6615.58 | 🔴 +0.06% |
| http_req_duration.avg | 60.73 | 59.83 | 🟢 -1.47% |
| http_req_duration.p(95) | 269.39 | 256.91 | 🟢 -4.63% |
| iteration_duration.avg | 6611.59 | 6615.64 | 🔴 +0.06% |
| data_received.count | 1,180,267,988 | 1,179,507,331 | 🟢 -0.06% |
| data_sent.count | 61,546,234 | 61,448,474 | 🟢 -0.16% |




### Result 

<details>



        script: test/k6/presence.ts

     scenarios: (100.00%) 1 scenario, 500 max VUs, 3m0s max duration (incl. graceful stop):
              * default: Up to 500 looping VUs for 2m30s over 3 stages (gracefulRampDown: 30s, gracefulStop: 30s)

  █ THRESHOLDS 

    active_clients_time
    ✓ 'p(95)<500' p(95)=113.84ms

    attach_documents_time
    ✓ 'p(95)<10000' p(95)=542.84ms

    deactivate_clients_time
    ✓ 'p(95)<10000' p(95)=111ms

    http_req_duration
    ✓ 'p(95)<10000' p(95)=256.91ms

    pushpulls_time
    ✓ 'p(95)<1000' p(95)=268ms

    transaction_success_rate
    ✓ 'rate>0.99' rate=100.00%

    transaction_time
    ✓ 'p(95)<30000' p(95)=7.4s

  █ TOTAL RESULTS 

    checks_total.......: 130624  840.950252/s
    checks_succeeded...: 100.00% 130624 out of 130624
    checks_failed......: 0.00%   0 out of 130624

    ✓ status is 200
    ✓ response has valid body

    CUSTOM
    active_clients....................: 8164    52.559391/s
    active_clients_success_rate.......: 100.00% 8164 out of 8164
    active_clients_time...............: avg=25.91ms  min=1ms      med=8ms    max=496ms    p(90)=74ms     p(95)=113.84ms
    attach_documents..................: 8164    52.559391/s
    attach_documents_success_rate.....: 100.00% 8164 out of 8164
    attach_documents_time.............: avg=150.67ms min=3ms      med=72ms   max=988ms    p(90)=455ms    p(95)=542.84ms
    deactivate_clients................: 8164    52.559391/s
    deactivate_clients_success_rate...: 100.00% 8164 out of 8164
    deactivate_clients_time...........: avg=21.59ms  min=0s       med=3ms    max=559ms    p(90)=63ms     p(95)=111ms   
    pushpulls.........................: 40820   262.796954/s
    pushpulls_success_rate............: 100.00% 40820 out of 40820
    pushpulls_time....................: avg=68.48ms  min=1ms      med=28ms   max=764ms    p(90)=195ms    p(95)=268ms   
    transaction_success_rate..........: 100.00% 8164 out of 8164
    transaction_time..................: avg=6.61s    min=6.01s    med=6.63s  max=8.8s     p(90)=7.2s     p(95)=7.4s    

    HTTP
    http_req_duration.................: avg=59.83ms  min=179.99µs med=15.6ms max=986.23ms p(90)=176.09ms p(95)=256.91ms
      { expected_response:true }......: avg=59.83ms  min=179.99µs med=15.6ms max=986.23ms p(90)=176.09ms p(95)=256.91ms
    http_req_failed...................: 0.00%   0 out of 65312
    http_reqs.........................: 65312   420.475126/s

    EXECUTION
    iteration_duration................: avg=6.61s    min=6.01s    med=6.63s  max=8.8s     p(90)=7.2s     p(95)=7.4s    
    iterations........................: 8164    52.559391/s
    vus...............................: 5       min=5              max=500
    vus_max...........................: 500     min=500            max=500

    NETWORK
    data_received.....................: 1.2 GB  7.6 MB/s
    data_sent.........................: 61 MB   396 kB/s



</details>

raararaara committed:
Revise the codes

raararaara committed:
Revise the codes

raararaara committed:
Merge remote-tracking branch 'origin/clusternodes' into clusternodes

raararaara commented on server/backend/database/mongo/client.go:
```diff
@@ -288,13 +316,55 @@
+				},
+			},
+		},
+		options.Find().SetSort(bson.D{
+			{Key: "is_leader", Value: -1},
+			{Key: "updated_at", Value: -1},
+		}),
+	)
```
Do I really need to introduce an index for this? An alternative might be to collect the results from the cursor first and then perform the sort on our side, rather than relying on MongoDB’s built-in sort.

raararaara replied:
If the data size is small, the cost of an unindexed sort may not be significant. Would it be acceptable to resolve this as is?

raararaara committed:
Apply coderabbit's suggestion

raararaara committed:
Apply coderabbit's suggestion

raararaara committed:
Apply coderabbit's suggestion

raararaara committed:
Revise the codes

raararaara committed:
Fix some errors

raararaara committed:
Replace hostname to rpcAddr

raararaara committed:
Cleanup the codes

raararaara committed:
Add related tests

raararaara committed:
Cleanup the codes

raararaara committed:
Fix test condition to ensure clear database

raararaara committed:
Add test for handling concurrent connection

hackerwins committed:
Increase cache hit rate with refined logging and DB fixes (#1482)

This commit primarily increases the client cache hit rate by ensuring
that cache entries are updated immediately after writes, which reduces
stale reads and improves overall consistency.

In the process of implementing this, I also standardized RPC logging
with clear levels for both successes and errors, fixed imprecise
“schema not found” errors, and unified the handling of missing records
to prevent edge-case failures.

raararaara committed:
Handle race condition

raararaara committed:
Merge branch 'main' into clusternodes

raararaara committed:
Fix error
